File: src/filesystem/docs/tools/add-command.ts
import type { PageMeta } from '../../page-types';

export const meta: PageMeta = {
  name: 'add-command',
  description: 'Install a custom terminal command into the portfolio site.',
  tech: ['Python', 'TypeScript'],
  sections: [
    {
      title: 'Overview',
      body: '<p><code>add_command.py</code> takes a <code>.ts</code> or <code>.js</code> file that exports a <code>Command</code> object and wires it into <code>src/commands/index.ts</code>, making the command available in the terminal immediately after the next build.</p>',
    },
    {
      title: 'Usage',
      body: `<pre><code>python tools/add_command.py &lt;command-file&gt; [--dry-run]
</code></pre>
<p>Run from the project root.</p>`,
    },
    {
      title: 'Arguments',
      body: `<table>
<thead>
<tr>
<th>Argument</th>
<th>Required</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>command-file</code></td>
<td>yes</td>
<td>Path to a <code>.ts</code> or <code>.js</code> file exporting a default <code>Command</code> object</td>
</tr>
<tr>
<td><code>--dry-run</code></td>
<td>no</td>
<td>Print generated output without writing any files</td>
</tr>
</tbody>
</table>`,
    },
    {
      title: 'Command File Format',
      body: `<p>The file must export a default object matching the <code>Command</code> interface. The filename stem becomes the command name (e.g. <code>ping.ts</code> → <code>ping</code>).</p>
<pre><code class="language-typescript">import type { Command } from './types';

const ping: Command = {
  description: 'Ping the home server',
  execute: async ({ args, addLine }) =&gt; {
    addLine('Pinging...');
    try {
      const res = await fetch('https://example.com/ping');
      addLine(res.ok ? 'Server is up.' : 'Server returned ' + res.status, res.ok ? 'output' : 'error');
    } catch {
      addLine('Could not reach server.', 'error');
    }
  },
};

export default ping;
</code></pre>
<p>The file is <strong>copied as-is</strong> into <code>src/commands/</code>. Any imports must already be relative to that directory (e.g. <code>import type { Command } from './types';</code>).</p>`,
    },
    {
      title: 'Command Naming',
      body: '<p>The stem of the filename is used as the command name and must match <code>^[a-z][a-z0-9_]*$</code> — lowercase letters, digits, and underscores only (e.g. <code>ping</code>, <code>check_status</code>, <code>roll</code>).</p>',
    },
    {
      title: 'Protected Commands',
      body: `<p>The following built-in commands cannot be overwritten:</p>
<p><code>help</code>, <code>whoami</code>, <code>ls</code>, <code>cd</code>, <code>cat</code>, <code>echo</code>, <code>clear</code>, <code>animations</code></p>`,
    },
    {
      title: 'What Gets Written',
      body: `<ul>
<li><code>src/commands/&lt;name&gt;.ts</code> — copied from source</li>
<li><code>src/commands/index.ts</code> — updated with import and registration</li>
</ul>`,
    },
    {
      title: 'Examples',
      body: `<pre><code class="language-bash">python tools/add_command.py tools/example-content/commands/ping.ts
python tools/add_command.py my-command.ts --dry-run
</code></pre>`,
    },
    {
      title: 'Notes',
      body: `<ul>
<li>If <code>&lt;name&gt;.ts</code> already exists in <code>src/commands/</code> it will be overwritten with a warning.</li>
<li>After writing, run <code>npm run lint</code> to verify the generated TypeScript.</li>
<li>To remove a command later, use <code>remove_command.py</code>.</li>
</ul>`,
    },
  ],
};

export const readme = `# add-command

${meta.description}

tech:  ${(meta.tech ?? []).join(' · ')}
`;

File: tools/add_command.py
#!/usr/bin/env python3
"""
add_command.py — Install a custom terminal command into the portfolio site.

Takes a .ts or .js file that exports a Command object and wires it into
src/commands/index.ts.

USAGE
    python tools/add_command.py <command-file> [--dry-run]

ARGUMENTS
    command-file   Path to a .ts or .js file exporting a default Command object.
                   The filename stem becomes the command name (e.g. ping.ts → ping).

COMMAND FILE FORMAT
    The file must export a default object matching the Command interface:

        import type { Command } from './types';

        const ping: Command = {
          description: 'Ping the home server',
          execute: async ({ args, addLine }) => {
            addLine('Pinging...');
            try {
              const res = await fetch('https://example.com/ping');
              addLine(res.ok ? 'Server is up.' : 'Server returned ' + res.status, res.ok ? 'output' : 'error');
            } catch {
              addLine('Could not reach server.', 'error');
            }
          },
        };

        export default ping;

    The file is copied as-is into src/commands/. Any imports must already be
    relative to that directory (e.g. import type { Command } from './types';).

COMMAND NAMING
    The stem of the filename is used as the command name and must be a valid
    JavaScript identifier: lowercase letters, digits, and underscores only
    (e.g. ping, check_status, roll).
"""

import argparse
import re
import shutil
import sys
from pathlib import Path

from _lib import insert_after_last_import

ROOT         = Path(__file__).parent.parent
COMMANDS_DIR = ROOT / "src" / "commands"
INDEX_FILE   = COMMANDS_DIR / "index.ts"

VALID_NAME = re.compile(r"^[a-z][a-z0-9_]*$")
PROTECTED  = {"help", "whoami", "ls", "cd", "cat", "echo", "clear", "animations"}


def update_index(name: str, src_stem: str, dry_run: bool) -> None:
    content = INDEX_FILE.read_text()

    import_line = f"import {name} from './{src_stem}';"
    if import_line in content:
        print(f"  [skip] '{name}' already registered in index.ts")
        return

    content = insert_after_last_import(content, import_line)

    content = re.sub(
        r"(\n)(};\n\nexport default commands;)",
        rf"\n  {name},\1\2",
        content,
    )

    if dry_run:
        print(f"\n--- commands/index.ts (updated) ---\n{content}")
    else:
        INDEX_FILE.write_text(content)
        print(f"  [updated] src/commands/index.ts")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Install a custom terminal command into the portfolio site.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("command_file", help="Path to the .ts or .js command file")
    parser.add_argument("--dry-run", action="store_true",
                        help="Print output without writing any files")
    args = parser.parse_args()

    src = Path(args.command_file)
    if not src.exists():
        print(f"ERROR: file not found: {src}")
        sys.exit(1)

    if src.suffix not in (".ts", ".js"):
        print(f"ERROR: file must be .ts or .js, got: {src.suffix}")
        sys.exit(1)

    name = src.stem
    if not VALID_NAME.match(name):
        print(f"ERROR: command name '{name}' must be lowercase letters, digits, "
              f"and underscores only (e.g. ping, check_status)")
        sys.exit(1)

    if name in PROTECTED:
        print(f"ERROR: '{name}' is a built-in command and cannot be overwritten.")
        sys.exit(1)

    dest = COMMANDS_DIR / src.name
    if dest.exists() and not args.dry_run:
        print(f"  WARNING: {dest.name} already exists — overwriting.")

    print(f"\nInstalling command '{name}'")
    print(f"  Source:  {src}")
    print(f"  Dest:    src/commands/{src.name}")

    if args.dry_run:
        print(f"\n--- {src.name} (copy of input) ---")
        print(src.read_text())
    else:
        shutil.copy2(src, dest)
        print(f"  [copied] src/commands/{src.name}")

    update_index(name, src.stem, args.dry_run)

    if not args.dry_run:
        print(f"\nDone. Run 'npm run lint' to verify.")


if __name__ == "__main__":
    main()
